File: src/main/java/pl/mjedynak/idea/plugins/builder/factory/AbstractPopupListFactory.java
package pl.mjedynak.idea.plugins.builder.factory;

import javax.swing.JList;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;
import pl.mjedynak.idea.plugins.builder.renderer.ActionCellRenderer;

public abstract class AbstractPopupListFactory {
    private ActionCellRenderer actionCellRenderer;

    @NotNull
    public JList<?> getPopupList() {
        JList<?> list = createList();
        list.setCellRenderer(cellRenderer());
        return list;
    }

    @Contract(value = "->new")
    @NotNull
    protected abstract JList<?> createList();

    private @NotNull ActionCellRenderer cellRenderer() {
        if (actionCellRenderer == null) {
            actionCellRenderer = new ActionCellRenderer();
        }
        return actionCellRenderer;
    }
}

File: src/main/kotlin/io/github/kingg22/kobra/builder/factory/AbstractPopupListFactory.kt
package io.github.kingg22.kobra.builder.factory

import io.github.kingg22.kobra.builder.renderer.ActionCellRenderer
import org.jetbrains.annotations.Contract
import org.jetbrains.annotations.VisibleForTesting
import javax.swing.JList

public abstract class AbstractPopupListFactory {
    @get:VisibleForTesting
    public var actionCellRenderer: ActionCellRenderer? = null
        private set

    public val popupList: JList<*> get() {
        val list = createList()
        list.setCellRenderer(cellRenderer())
        return list
    }

    @Contract(value = "->new")
    protected abstract fun createList(): JList<*>

    private fun cellRenderer(): ActionCellRenderer {
        if (actionCellRenderer == null) {
            actionCellRenderer = ActionCellRenderer()
        }
        return actionCellRenderer!!
    }
}
